provider/framework_provider.go

```diff
@@ -0,0 +1,212 @@
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"os"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/provider"
+	"github.com/hashicorp/terraform-plugin-framework/provider/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+// Ensure redisCloudProvider satisfies provider interface.
+var _ provider.Provider = &redisCloudProvider{}
+
+// redisCloudProvider defines the provider implementation.
+type redisCloudProvider struct {
+	// version is set to the provider version on release, "dev" when the
+	// provider is built and ran locally, and "test" when running acceptance
+	// testing.
+	version string
+}
+
+// redisCloudProviderModel describes the provider data model.
+type redisCloudProviderModel struct {
+	Url       types.String `tfsdk:"url"`
+	ApiKey    types.String `tfsdk:"api_key"`
+	SecretKey types.String `tfsdk:"secret_key"`
+}
+
+func (p *redisCloudProvider) Metadata(_ context.Context, _ provider.MetadataRequest, _ *provider.MetadataResponse) {
+}
+
+func (p *redisCloudProvider) Schema(_ context.Context, _ provider.SchemaRequest, resp *provider.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Attributes: map[string]schema.Attribute{
+			"url": schema.StringAttribute{
+				MarkdownDescription: fmt.Sprintf("This is the URL of Redis Cloud and will default to `https://api.redislabs.com/v1`. This can also be set by the `%s` environment variable.", RedisCloudUrlEnvVar),
+				Optional:            true,
+			},
+			"api_key": schema.StringAttribute{
+				MarkdownDescription: fmt.Sprintf("This is the Redis Cloud API key. It must be provided but can also be set by the `%s` environment variable.", rediscloudApi.AccessKeyEnvVar),
+				Optional:            true,
+			},
+			"secret_key": schema.StringAttribute{
+				MarkdownDescription: fmt.Sprintf("This is the Redis Cloud API secret key. It must be provided but can also be set by the `%s` environment variable.", rediscloudApi.SecretKeyEnvVar),
+				Optional:            true,
+			},
+		},
+		Blocks: map[string]schema.Block{},
+	}
+}
+
+func (p *redisCloudProvider) Configure(ctx context.Context, req provider.ConfigureRequest, resp *provider.ConfigureResponse) {
+	tflog.Info(ctx, "Configuring Rediscloud client")
+
+	var config []rediscloudApi.Option
+	ua := fmt.Sprintf("Terraform/%s (+https://www.terraform.io) Terraform-Plugin-Framework terraform-provider-rediscloud/%s", req.TerraformVersion, p.version)
+	config = append(config, rediscloudApi.AdditionalUserAgent(ua))
+
+	// Retrieve provider data from configuration
+	var data redisCloudProviderModel
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// All three values should be known before any other resource is applied
+	if data.Url.IsUnknown() {
```

Comment:
I think IsUnknown returns true if the attribute wasn't defined in the provider block within the HCL, so this block - and the similar ones below - would return an error rather than use environment variables

Reply:
That's not how [this documentation](https://developer.hashicorp.com/terraform/plugin/framework/handling-data/attributes#null-and-unknown-values) reads to me, but I'll have a play around with it.

Particularly this bit:

> Because they can result from interpolations in the practitioner's config, you have no control over what attributes may contain an unknown value. However, by the time a resource is expected to be created, read, updated, or deleted, only its computed attributes can be unknown. The rest are guaranteed to have known values (or be null).

Provider configuration values can be unknown, and providers should handle that situation, even if that means just returning an error.